src/chunk_append/planner.c

```diff
@@ -0,0 +1,326 @@
+
+					foreach (lc, clauses)
+					{
+						Node *clause = (Node *) transform_restrict_info_clause(
+							castNode(RestrictInfo, lfirst(lc))->clause);
+						clause = adjust_appendrel_attrs_compat(root, clause, appinfo);
+						chunk_clauses = lappend(chunk_clauses, clause);
+					}
+					chunk_ri_clauses = lappend(chunk_ri_clauses, chunk_clauses);
+					break;
+				}
+				default:
+					elog(ERROR, "invalid child of chunk append: %u", nodeTag(plan));
+					break;
+			}
+		}
+		Assert(list_length(cscan->custom_plans) == list_length(chunk_ri_clauses));
+	}
+
+	if (root->limit_tuples > 0 && root->limit_tuples <= PG_UINT32_MAX)
+		limit = root->limit_tuples;
+
+	custom_private =
+		lappend(custom_private, list_make2_oid(((ChunkAppendPath *) path)->ht_reloid, limit));
+	custom_private = lappend(custom_private, chunk_ri_clauses);
+
+	cscan->custom_private = custom_private;
+
+	return &cscan->scan.plan;
+}
+
+/*
+ * make_sort --- basic routine to build a Sort plan node
+ *
+ * Caller must have built the sortColIdx, sortOperators, collations, and
+ * nullsFirst arrays already.
+ */
+static Sort *
+make_sort(Plan *lefttree, int numCols, AttrNumber *sortColIdx, Oid *sortOperators, Oid *collations,
+		  bool *nullsFirst)
+{
+	Sort *node = makeNode(Sort);
+	Plan *plan = &node->plan;
+
+	plan->targetlist = lefttree->targetlist;
+	plan->qual = NIL;
+	plan->lefttree = lefttree;
+	plan->righttree = NULL;
+	node->numCols = numCols;
+	node->sortColIdx = sortColIdx;
+	node->sortOperators = sortOperators;
+	node->collations = collations;
+	node->nullsFirst = nullsFirst;
+
+	return node;
+}
+
+static AppendRelInfo *
+get_appendrelinfo(PlannerInfo *root, Index rti)
+{
+#if PG96 || PG10
+	ListCell *lc;
+	foreach (lc, root->append_rel_list)
+	{
+		AppendRelInfo *appinfo = lfirst(lc);
+		if (appinfo->child_relid == rti)
+			return appinfo;
+	}
+#else
+	if (root->append_rel_array[rti])
+		return root->append_rel_array[rti];
+#endif
+	ereport(ERROR,
+			(errcode(ERRCODE_INTERNAL_ERROR), errmsg("no appendrelinfo found for index %d", rti)));
+	pg_unreachable();
+}
+
+#define DATATYPE_PAIR(left, right, type1, type2)                                                   \
+	((left == type1 && right == type2) || (left == type2 && right == type1))
+
+/*
+ * Cross datatype comparisons between DATE/TIMESTAMP/TIMESTAMPTZ
+ * are not immutable which prevents their usage for chunk exclusion.
+ * Unfortunately estimate_expression_value will not estimate those
+ * expressions which makes them unusable for execution time chunk
+ * exclusion with constraint aware append.
+ * To circumvent this we inject casts and use an operator
+ * with the same datatype on both sides when constifying
+ * restrictinfo. This allows estimate_expression_value
+ * to evaluate those expressions and makes them accessible for
+ * execution time chunk exclusion.
+ *
+ * The following transformations are done:
+ * TIMESTAMP OP TIMESTAMPTZ => TIMESTAMP OP (TIMESTAMPTZ::TIMESTAMP)
```

Comment:
Out of curiosity how come `TIMESTAMPTZ::TIMESTAMP` instead of `TIMESTAMP::TIMESTAMPTZ`?

Reply:
the assumption is left side is the column so we wanna have all the expressions on the right so it can be evaluated